stan/math/prim/prob/hmm_marginal_lpdf.hpp

```diff
@@ -0,0 +1,187 @@
+ * and parameters theta, return the log marginal density, log
+ * p(y | theta). In this setting, the hidden states are discrete
+ * and take values over the finite space {1, ..., K}.
+ * The marginal lpdf is obtained via a forward pass, and
+ * the derivative is calculated with an adjoint method,
+ * see (Betancourt, Margossian, & Leos-Barajas, 2020).
+ *
+ * @tparam T_omega type of the log likelihood matrix
+ * @tparam T_Gamma type of the transition matrix
+ * @tparam T_rho type of the initial guess vector
+ *
+ * @param[in] log_omega log matrix of observational densities.
+ *              The (i, j)th entry corresponds to the
+ *              density of the ith observation, y_i,
+ *              given x_i = j.
+ * @param[in] Gamma transition density between hidden states.
+ *              The (i, j)th entry is the probability that x_n = j,
+ *              given x_{n - 1} = i. The rows of Gamma are simplexes.
+ * @param[in] rho initial state
+ * @throw `std::domain_error` if Gamma is not square.
+ * @throw `std::invalid_argument` if the size of rho is not
+ * the number of rows of Gamma.
+ * @throw `std::domain_error` if rho is not a simplex.
+ * @throw `std::domain_error` if the rows of Gamma are
+ * not a simplex.
+ * @return log marginal density.
+ */
+template <typename T_omega, typename T_Gamma, typename T_rho>
+inline auto hmm_marginal_lpdf(
+    const Eigen::Matrix<T_omega, Eigen::Dynamic, Eigen::Dynamic>& log_omegas,
+    const Eigen::Matrix<T_Gamma, Eigen::Dynamic, Eigen::Dynamic>& Gamma,
+    const Eigen::Matrix<T_rho, Eigen::Dynamic, 1>& rho) {
+  using T_partial_type = partials_return_t<T_omega, T_Gamma, T_rho>;
+  using eig_matrix_partial
+      = Eigen::Matrix<T_partial_type, Eigen::Dynamic, Eigen::Dynamic>;
+  using eig_vector_partial = Eigen::Matrix<T_partial_type, Eigen::Dynamic, 1>;
+  int n_states = log_omegas.rows();
+  int n_transitions = log_omegas.cols() - 1;
+
+  check_square("hmm_marginal_lpdf", "Gamma", Gamma);
+  check_consistent_size("hmm_marginal_lpdf", "Gamma", row(Gamma, 1), n_states);
+  check_consistent_size("hmm_marginal_lpdf", "rho", rho, n_states);
+  check_simplex("hmm_marginal_lpdf", "rho", rho);
+  for (int i = 0; i < Gamma.rows(); ++i) {
+    check_simplex("hmm_marginal_lpdf", "Gamma[i, ]", row(Gamma, i + 1));
+  }
+
+  operands_and_partials<Eigen::Matrix<T_omega, Eigen::Dynamic, Eigen::Dynamic>,
+                        Eigen::Matrix<T_Gamma, Eigen::Dynamic, Eigen::Dynamic>,
+                        Eigen::Matrix<T_rho, Eigen::Dynamic, 1> >
+      ops_partials(log_omegas, Gamma, rho);
+
+  eig_matrix_partial alphas(n_states, n_transitions + 1);
+  eig_vector_partial alpha_log_norms(n_transitions + 1);
+  eig_matrix_partial omegas;
+  auto Gamma_val = value_of(Gamma).eval();
+
+  // compute the density using the forward algorithm.
+  {
+    const auto log_omegas_val = value_of(log_omegas).eval();
+    const auto rho_val = value_of(rho).eval();
+    omegas = log_omegas_val.array().exp();
```

Comment:
- The constants are a gift from our friend @SteveBronder but I'm with you that we can remove them.
- move the declaration of omega
- also removed log_omegas_val, since it does not get used.